fix(bug): Make table and button consistent in settings and certification page.

nayabatir1 opened the pull request:
Affected page
- certification page
- settings page

Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [X] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [X] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [X] My pull request targets the `main` branch of freeCodeCamp.
- [X] I have tested these changes either locally on my machine, or GitPod.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Closes #45949 

<!-- Feel free to add any additional description of changes below this line -->
Affected pages
- settings
- certification

![image](https://user-images.githubusercontent.com/91016903/197358895-4422b7b4-4a71-4559-a594-f79f76aab486.png)
![image](https://user-images.githubusercontent.com/91016903/197358906-fc2f4552-e13e-480d-8bdc-1649f9b9fca9.png)

nayabatir1 committed:
fix(bug): Make table and button consistent

Affected page
- certification page
- settings page

ahmadabdolsaheb commented:
Thanks for the pr. 
Quick notes:
- The buttons should say "View", and there should be <th> elements for "Project name", "Solution", and "Completed"
- Only Timeline and Certification tables would have condensed={true} striped={true}

nayabatir1 committed:
fix(bug): changed certification button text to View

nayabatir1 committed:
fix(bug): Timeline table width increased

nayabatir1 commented:
For long challenges with long title, it's content are wrapped. So I've also increased table width.

### Before
![image](https://user-images.githubusercontent.com/91016903/197623347-e5ad2d3e-74da-44d0-9223-e63706136126.png)


### After
![image](https://user-images.githubusercontent.com/91016903/197623205-b47dac76-5b95-403f-9d95-99f53914f475.png)

naomi-lgbt reviewed (dismissed):
banger pr, thanks for contributing!

Sboonny commented on client/src/client-only-routes/show-project-links.tsx:
```diff
@@ -90,6 +91,7 @@ const ShowProjectLinks = (props: ShowProjectLinksProps): JSX.Element => {
   };
 
   const renderProjectsFor = (certName: string) => {
+    console.log(certName);
```
Stray log
```suggestion
```

Sboonny commented on client/src/client-only-routes/show-project-links.tsx:
```diff
@@ -160,7 +161,10 @@ const ShowProjectLinks = (props: ShowProjectLinksProps): JSX.Element => {
       <Spacer />
-      <ul>{renderProjectsFor(certName)}</ul>
+      <Table striped>
+        <thead></thead>
```
```suggestion
````

How is this used and is it beneficial for you to remove them?

Sboonny commented on client/src/components/profile/components/time-line.tsx:
```diff
@@ -1,6 +1,6 @@
 import { Button, Modal, Table } from '@freecodecamp/react-bootstrap';
 import Loadable from '@loadable/component';
-import { useStaticQuery, graphql } from 'gatsby';
+import { graphql, useStaticQuery } from 'gatsby';
```
Tf why is graphql in gatspy. 🤣

nayabatir1 replied:
🤷

Sboonny commented on client/src/components/profile/components/time-line.tsx:
```diff
@@ -165,62 +165,60 @@ function TimelineInner({
+          </thead>
+          <tbody>
+            {sortedTimeline.slice(startIndex, endIndex).map(renderCompletion)}
+          </tbody>
+        </Table>
+      )}
+      {id && (
+        <Modal
+          aria-labelledby='contained-modal-title'
+          onHide={closeSolution}
+          show={solutionOpen}
+        >
+          <Modal.Header closeButton={true}>
+            <Modal.Title id='contained-modal-title' className='text-center'>
+              {`${username}'s Solution to ${
+                idToNameMap.get(id)?.challengeTitle ?? ''
+              }`}
+            </Modal.Title>
+          </Modal.Header>
+          <Modal.Body>
+            <SolutionViewer
+              challengeFiles={challengeData.challengeFiles}
+              solution={challengeData.solution ?? ''}
+            />
+          </Modal.Body>
+          <Modal.Footer>
+            <Button onClick={closeSolution}>{t('buttons.close')}</Button>
+          </Modal.Footer>
+        </Modal>
+      )}
+      {totalPages > 1 && (
+        <TimelinePagination
+          firstPage={firstPage}
+          lastPage={lastPage}
+          nextPage={nextPage}
+          pageNo={pageNo}
+          prevPage={prevPage}
+          totalPages={totalPages}
+        />
+      )}
```
> there should be elements for "Project name", "Solution", and "Completed"

How did you address this, btw? because I think I am misunderstanding it as adding `<span class='sr-only'>Project name</span>`. :thinking:

nayabatir1 replied:
The only thing I did in this file is increase the table width as all three columns were already there line:184. 
You can check `renderCompletion` function.

Sboonny replied:
So I was misunderstanding, thank you for conforming

https://github.com/freeCodeCamp/freeCodeCamp/blob/898a9331a6b5889f2a4122f379b2022a73aae4ea/client/src/components/profile/components/time-line.tsx#L118-L150=plain?1

Sboonny reviewed (commented):
Small nitpicks and one question about Ahmad previous comment https://github.com/freeCodeCamp/freeCodeCamp/pull/48211#issuecomment-1288818880

nayabatir1 committed:
Update client/src/client-only-routes/show-project-links.tsx

Co-authored-by: Muhammed Mustafa <muhammed@freecodecamp.org>

nayabatir1 committed:
Update client/src/client-only-routes/show-project-links.tsx

Co-authored-by: Muhammed Mustafa <muhammed@freecodecamp.org>

moT01 reviewed (dismissed):
I gave this some testing, it’s looking good @nayabatir1. Here's everything I found that I think is specifically related to this PR...

-One main thing I noticed is that this doesn't appear to create a common component, [as described here](https://github.com/freeCodeCamp/freeCodeCamp/issues/45949#issuecomment-1265138478), that all the tables use. I don't think that's a big issue - although, I think it would have been nice and simplified things a bit. I would be fine without it.

-On the certifications, the project name is not vertically centered in the table row:
![Screen Shot 2022-10-31 at 4 43 14 PM](https://user-images.githubusercontent.com/20648924/199116218-f8550ced-9ebe-4e80-b3e6-330a9c2907ac.png)

-On the settings page, the table has stripes - it was [mentioned above](https://github.com/freeCodeCamp/freeCodeCamp/pull/48211#issuecomment-1288818880) that only the other two places should have stripes. I think that's what that means, anyway. So, I think we want to remove them there. Can you confirm @ahmadabdolsaheb.

I think if we fixed those second two things, we could probably get this in - however...

<hr>

Here’s some other things I noticed that were likely pre-existing and not from this PR or specifically part of the issue - but I’m going to bring them up in case we want to fix any of them here.

-Probably common on all the tables, but most noticeable on the profile timeline - the view button (solution and completed column) have different widths depending on the length of the title. Sometimes quite skinny, and other times, very wide. I would like to see those two columns have a fixed width. Something that might have been easier to fix with that common component. 

-On the JavaScript certification, the links use "solution" text instead of view button:
![Screen Shot 2022-10-31 at 4 54 38 PM](https://user-images.githubusercontent.com/20648924/199117822-3a72e488-40b6-4f04-9e8c-4647bff16509.png)
And actually, the titles there are vertically centered, so it must be the view button that is causing them not to be.

I gave this a rebase locally cause it didn't have the [icons added here](https://github.com/freeCodeCamp/freeCodeCamp/pull/48268) when I ran it. Noticed that the view button on the certifications don't have that icon.

<hr>

I guess one other thing related to this PR...
> For challenges with long title, its content are wrapped. So I've also increased the table width.

I don't really mind that, I think it looks good - but we have a pretty standard margin on most of the pages (not all), and this deviates from that. Maybe @ahmadabdolsaheb can weigh in on that, as well.

ahmadabdolsaheb commented:
Thanks for taking a look Tom.

> For challenges with long title, its content are wrapped. So I've also increased the table width.

That would not be necessary. They will eventually wrap if you view most of the titles on smaller screens. We would need to keep the spacing similar across tables.

> One main thing I noticed is that this doesn't appear to create a common component, https://github.com/freeCodeCamp/freeCodeCamp/issues/45949#issuecomment-1265138478, that all the tables use. I don't think that's a big issue - although, I think it would have been nice and simplified things a bit. I would be fine without it.

Having one single component will also help with migration to component library. So, it would be great to have it now.

> On the settings page, the table has stripes - it was https://github.com/freeCodeCamp/freeCodeCamp/pull/48211#issuecomment-1288818880 that only the other two places should have stripes. I think that's what that means, anyway. So, I think we want to remove them there. Can you confirm @ahmadabdolsaheb.

Settings page should not be condensed or striped, the other two tables should.

nayabatir1 committed:
fix(curriculum): remoeve stripe props from certification table

nayabatir1 committed:
fix(curriculum): time line table width reverted to original

nayabatir1 commented:
> That would not be necessary. They will eventually wrap if you view most of the titles on smaller screens. We would need to keep the spacing similar across tables.

I've reverted time-line table width to original.

> Having one single component will also help with migration to component library. So, it would be great to have it now.

I did plan to create table component but we are using bootstrap `Table` which is itself a component and each page has their own header and body. So, how are we planning to make it reusable.

> Settings page should not be condensed or striped, the other two tables should.

I've removed stripe props from settings table.

ahmadabdolsaheb commented:
Thanks for implementing the changes. As you suggested all tables do not need to be in one table component.

The certification would need a <thead>, and 'solution' need to be a 'View' button similar to the timeline.

The project titles should not be bold on the certifications. Basically, similar elements should look similar on all tables.

Finally, we do have 'View' translated on buttons.view. As a result, adding a new key is not necessary.

nayabatir1 committed:
fix(client): use buttons.view from translation.json

nayabatir1 committed:
fix(client): use normal font weight for project title

ahmadabdolsaheb commented:
You would need table header elements, here is an example:
```js
            <thead>
              <tr>
                <th>{t('profile.challenge')}</th>
                <th>{t('settings.labels.solution')}</th>
```

nayabatir1 commented:
![image](https://user-images.githubusercontent.com/91016903/200389443-53d046af-b584-40ef-879e-5c5fa2be6868.png)
For above links shall I convert it to button too?
Also shall I add another column for source ?

ahmadabdolsaheb commented:
> ![image](https://user-images.githubusercontent.com/91016903/200389443-53d046af-b584-40ef-879e-5c5fa2be6868.png) For above links shall I convert it to button too?

Yes, those are gonna become a view dropdown button. Here is an example we are using in settings: 

<img width="956" alt="Screen Shot 2022-11-15 at 10 13 33 AM" src="https://user-images.githubusercontent.com/4591597/201853795-a1d7ca84-421c-43de-bb70-ca825ba787f3.png">

nayabatir1 committed:
fix(client): certification page table button

Sboonny commented on client/src/components/solution-display-widget/index.tsx:
```diff
@@ -30,34 +30,44 @@ export function SolutionDisplayWidget({
+          id={`dropdown-for-${id}`}
+          title={t('buttons.view')}
+        >
+          <MenuItem bsStyle='primary'>
+            <a href={solution ?? ''} rel='noopener noreferrer' target='_blank'>
+              {t('certification.project.solution')}
+            </a>
+          </MenuItem>
+          <MenuItem bsStyle='primary'>
+            <a href={githubLink} rel='noopener noreferrer' target='_blank'>
+              {t('certification.project.source')}
+            </a>
+          </MenuItem>
```
`MenuItem` is an anchor element, I hope. So you don't need to add an anchor element as its child, you can move the attributes to `MenuItem`

Sboonny replied:
because menuItem doesn't have href prop in it, it has button role which we don't want as it may confuse screen reader.

nayabatir1 replied:
Just to summarize
I've to modify menuItem component to accept href prop too and remove anchor tag child correct?

Sboonny replied:
I don't think you need to modify menuItem because it accepts the href prop. But yes, please remove the anchor tag child 👍 

https://github.com/freeCodeCamp/freeCodeCamp/blob/4e14899e1ce3f1b6f3ef4a5deb24240066b4baef/client/src/components/solution-display-widget/index.tsx#L115-L130

Sboonny commented on client/src/components/solution-display-widget/index.tsx:
```diff
@@ -156,5 +166,7 @@ export function SolutionDisplayWidget({
           none: MissingSolutionComponent
         };
 
+  console.log(getSolutionDisplayType(completedChallenge));
+
```
```suggestion
```

nayabatir1 committed:
fix(client): remove child anchor tag from menuItem

ahmadabdolsaheb commented:
> ![image](https://user-images.githubusercontent.com/91016903/202020463-7d82ed8c-595e-42a8-90d5-8676bd88c776.png)
> 
> @ahmadabdolsaheb in certification page dropdown width is not 100% of column width, though I've reused dropdown button code. Let me know if you prefer width to be 100%

The drop-downs should look consistant across the UI. Yes it should take full width.

nayabatir1 committed:
fix(client): updated cerification page button width

Sboonny commented on client/src/components/solution-display-widget/index.tsx:
```diff
@@ -7,7 +7,7 @@ import React from 'react';
 import { CompletedChallenge } from '../../redux/prop-types';
 import { getSolutionDisplayType } from '../../utils/solution-display-type';
-
+import './style.css';
```
```suggestion
```

Sboonny commented on client/src/components/solution-display-widget/index.tsx:
```diff
@@ -35,32 +35,30 @@ export function SolutionDisplayWidget({
+    <DropdownButton
+      block={true}
+      bsStyle='primary w-100'
+      className='btn-invert'
```
```suggestion
      className='btn-invert col-xs-12'
```

Sboonny commented on client/src/components/solution-display-widget/style.css:
```diff
@@ -0,0 +1,4 @@
+.w-100,
+.w-100 > .dropdown-menu {
+  width: 100%;
+}
```
```suggestion
```

nayabatir1 replied:
Hey I looked into your proposed solution `col-xs-12`, it didn't worked for `MenuItem`
`DropdownButton` and `MenuItem` both should have `width` of `100%`. Otherwise only `View` button will have full width and not `MenuItem`

using my custom css `w-100`
![image](https://user-images.githubusercontent.com/91016903/202535255-b4918dcb-13e3-4160-a867-795eb85ef131.png)


using `col-xs-12`
![image](https://user-images.githubusercontent.com/91016903/202535141-4584794d-dbb5-4981-ae76-cbc706862d10.png)

Which one you would prefer ?

Sboonny replied:
Thank you for checking it out, if it doesn't look right. Let me see if there is a class in the codebase there with the changes, that make our life easier to update all of them later.

nayabatir1 replied:
Sure, `.dropdown-menu` is the class of `ul` element which `MenuItem` adds dom. I'm using that class  to update its width to 100%

Sboonny replied:
Oh, I mean using the styles in `time-line.css`, as this component is doing the same thing and almost have the same style, we can use the same file

nayabatir1 replied:
from `timeline.css`
```css
.timeline-row .solutions-dropdown .dropdown {
  width: 100%;
}
```
It didn't worked as `.dropdown` don't have any parent element with class `solutions-dropdown` whose parent class is `timeline-row`

Sboonny replied:
True, adding `<div className='.solutions-dropdown'>` as parent gives width `100%`.

![image](https://user-images.githubusercontent.com/88248797/202775891-ebc501bc-0490-4107-a5a1-f12c724f1704.png)

and it will look the same as functions below it 

https://github.com/freeCodeCamp/freeCodeCamp/blob/6ad3a78db1fd6bf37a56db700fc33e3fb9cb424a/client/src/components/solution-display-widget/index.tsx#L90-L105

Sboonny replied:
I have been checking the wrong certification the whole time?

nayabatir1 replied:
Nope, adding parent element with class name as `.solutions-dropdown` didn't worked either. Check quality-assurance-v7 certification.

Sboonny reviewed (commented):
I have tried to use CSS module in a previous PR, too. https://github.com/freeCodeCamp/freeCodeCamp/pull/45964/files

But there is no need for it, we are still using bootstrap. className `col-xs-12` should make them alike

Sboonny reviewed (approved):
LGTM 🎉, thank you for your contribution to the page! 👍.

bbsmooth commented:
So of course I'm going to rain on the parade a little here. I apologize for not evaluating this sooner for accessibility. I didn't realize there was a drop down menu button.

Concerning that drop down View button, I would recommend we not use the menu roles on the list and list items. These are really just a list of links and thus no menu roles are needed (see our main Menu navigation). When a screen reader user hears that this is a menu then they expect it to act in a certain way and be composed of action items that do something, not merely links that open a new page. 

Really, this just needs to be a button with `aria-expanded` and then a plain list with no roles. You can keep the arrow navigation if you want but it is not necessary. The user should be able to tab through the list items though. And definitely keep the `Esc` key for closing the expanded list.

But if you want to keep this as an actual "menu" interface then there are a few things that need to be fixed. I will hold off on that until I know which way you want to proceed.

ahmadabdolsaheb commented:
Thank you @bbsmooth for the valuable insights. It seems like we would need to create a custom component with custom styles if we want to go with the simple button and list. 

Since the current dropdown menu is closest to what we need in the react-bootstrap component library, could you please let us know what needs to be fixed in order to make it more accessible?

Also, since the dropdown menu already exist on the settings tables, should we dedicate an issue to discuss its accessibility independently once this pr is in?

ahmadabdolsaheb commented:
sorry for the double tag @bbsmooth, just wanted to make sure this pr didn't get lost in the notifications :)

bbsmooth commented:
Hey @ahmadabdolsaheb, the current accessibility on these view buttons is not terrible, there are just a few things that could be improved regarding the functionality of the menu button itself and the overall accessibility of how they are being used in the table. It's probably best if I create separate issues for these things after this PR has been merged.

ahmaxed committed:
feat: relocate the styles

ahmaxed committed:
feat: clean up styles
